neo4j/session_with_context.go

```diff
@@ -545,49 +564,105 @@ func (s *sessionWithContext) getServers(mode idb.AccessMode) func() []string {
 	if err != nil {
 		return nil, errorutil.WrapError(err)
 	}
 
-	conn, err := s.pool.Borrow(
-		ctx,
-		s.getServers(mode),
-		timeout != 0,
-		s.config.BoltLogger,
-		livenessCheckTimeout,
-		s.auth)
+	conn, err := s.borrowConnection(ctx, serverList, livenessCheckTimeout)
 	if err != nil {
 		return nil, errorutil.WrapError(err)
 	}
 
+	// Return the connection if SSR is disabled, and we used the home database guess, then resolve as normal.
+	if !conn.IsSsrEnabled() && usedHomeDbGuess {
+		s.pool.Return(ctx, conn)
+
+		serverList, _, err = s.getServerList(ctx, mode)
+		if err != nil {
+			return nil, errorutil.WrapError(err)
+		}
+
+		conn, err = s.borrowConnection(ctx, serverList, livenessCheckTimeout)
+		if err != nil {
+			return nil, errorutil.WrapError(err)
+		}
+	}
+
+	conn.SetPinHomeDatabaseCallback(s.pinHomeDatabaseCallback)
+
 	// Select database on server
 	if s.config.DatabaseName != idb.DefaultDatabase {
-		dbSelector, ok := conn.(idb.DatabaseSelector)
-		if !ok {
-			s.pool.Return(ctx, conn)
-			return nil, &UsageError{Message: "Database does not support multi-database"}
+		if err := s.selectDatabase(ctx, conn); err != nil {
+			return nil, err
 		}
-		dbSelector.SelectDatabase(s.config.DatabaseName)
 	}
 
 	return conn, nil
 }
 
+// applyConnectionTimeout sets a timeout on the context if configured.
+func (s *sessionWithContext) applyConnectionTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
+	timeout := s.driverConfig.ConnectionAcquisitionTimeout
+	if timeout > 0 {
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		deadline, _ := ctx.Deadline()
+		s.log.Debugf(log.Session, s.logId, "connection acquisition timeout is %s, resolved deadline is: %s", timeout, deadline)
+		return ctx, cancel
+	}
+	if deadline, ok := ctx.Deadline(); ok {
+		s.log.Debugf(log.Session, s.logId, "connection acquisition user-provided deadline is: %s", deadline)
+	}
+	return ctx, nil
+}
+
+// getServerList resolves the server list based on the session configuration.
+// It returns a list of servers, a boolean indicating whether the home database guess was used, and an error if resolution fails.
+func (s *sessionWithContext) getServerList(ctx context.Context, mode idb.AccessMode) ([]string, bool, error) {
+	if s.config.DatabaseName != "" {
+		if err := s.resolveHomeDatabase(ctx); err != nil {
+			return nil, false, err
+		}
+		serverList, err := s.getOrUpdateServers(ctx, mode, false)
+		return serverList, false, err
+	}
+
+	// Use cached home database if available
+	if s.homeDbGuess != "" && s.router.GetTable(s.homeDbGuess) != nil && s.cache.IsEnabled() {
+		serverList, err := s.getOrUpdateServers(ctx, mode, true)
+		return serverList, true, err
+	}
+
+	// Resolve home database if cache is unavailable, there's no guess, or a table doesn't exist.
+	if err := s.resolveHomeDatabase(ctx); err != nil {
+		return nil, false, err
+	}
+	serverList, err := s.getOrUpdateServers(ctx, mode, s.cache.IsEnabled())
```

Comment:
```suggestion
	serverList, err := s.getOrUpdateServers(ctx, mode, false)
```
At this point, we've explicitly resolved and pinned the homeDB. We don't care about the cache any longer. It also didn't align with the return value below claiming that no cache was used.